caas/kubernetes/provider/k8s.go

```diff
@@ -1366,32 +1367,132 @@ type configMapNameFunc func(fileSetName string) string
 			podSpec.Containers[i].VolumeMounts = append(podSpec.Containers[i].VolumeMounts, core.VolumeMount{
-				Name:      cfgName,
+				// TODO(caas): add more config fields support(SubPath, ReadOnly, etc).
+				Name:      vol.Name,
 				MountPath: fileSet.MountPath,
 			})
 		}
 	}
 	return nil
 }
 
+// pushUniqVolume ensures to only add unique volumes because k8s will not schedule pods if it has deplucated volumes.
+func pushUniqVolume(podSpec *core.PodSpec, vol core.Volume) {
+	for _, v := range podSpec.Volumes {
+		if reflect.DeepEqual(v, vol) {
+			return
+		}
+	}
+	podSpec.Volumes = append(podSpec.Volumes, vol)
+}
+
+func (k *kubernetesClient) fileSetToVolume(
+	appName string,
+	annotations map[string]string,
+	workloadSpec *workloadSpec,
+	fileSet specs.FileSet,
+	cfgMapName configMapNameFunc,
+) (core.Volume, error) {
+	fileRefsToVolItems := func(fs []specs.FileRef) (out []core.KeyToPath) {
+		for _, f := range fs {
+			out = append(out, core.KeyToPath{
+				Key:  f.Key,
+				Path: f.Path,
+				Mode: f.Mode,
+			})
+		}
+		return out
+	}
+
+	vol := core.Volume{Name: fileSet.Name}
+	if len(fileSet.Files) > 0 {
+		vol.Name = cfgMapName(fileSet.Name)
+		if _, err := k.ensureConfigMapLegacy(filesetConfigMap(vol.Name, k.getConfigMapLabels(appName), annotations, &fileSet)); err != nil {
+			return vol, errors.Annotatef(err, "creating or updating ConfigMap for file set %v", vol.Name)
+		}
+		vol.ConfigMap = &core.ConfigMapVolumeSource{
+			LocalObjectReference: core.LocalObjectReference{
+				Name: vol.Name,
+			},
+		}
+		for _, f := range fileSet.Files {
+			vol.ConfigMap.Items = append(vol.ConfigMap.Items, core.KeyToPath{
+				Key:  f.Path,
+				Path: f.Path,
+				Mode: f.Mode,
+			})
+		}
+	} else if fileSet.HostPath != nil {
+		t := core.HostPathType(fileSet.HostPath.Type)
+		vol.HostPath = &core.HostPathVolumeSource{
+			Path: fileSet.HostPath.Path,
+			Type: &t,
+		}
+	} else if fileSet.EmptyDir != nil {
+		vol.EmptyDir = &core.EmptyDirVolumeSource{
+			Medium:    core.StorageMedium(fileSet.EmptyDir.Medium),
+			SizeLimit: fileSet.EmptyDir.SizeLimit,
+		}
+	} else if fileSet.ConfigMap != nil {
+		found := false
+		refName := fileSet.ConfigMap.Name
+		for cfgN := range workloadSpec.ConfigMaps {
+			if cfgN == refName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return vol, errors.NotValidf("non existing config map %q", refName)
+		}
+
+		vol.ConfigMap = &core.ConfigMapVolumeSource{
+			LocalObjectReference: core.LocalObjectReference{
+				Name: refName,
+			},
+			DefaultMode: fileSet.ConfigMap.DefaultMode,
+			Optional:    fileSet.ConfigMap.Optional,
+			Items:       fileRefsToVolItems(fileSet.ConfigMap.Files),
+		}
+	} else if fileSet.Secret != nil {
+		found := false
+		refName := fileSet.Secret.Name
+		for _, secret := range workloadSpec.Secrets {
+			if secret.Name == refName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return vol, errors.NotValidf("non existing secret %q", refName)
```

Comment:
NewNotValidf(nil, "cannot mount a volume using a secret if the secret %q is not specified in the pod spec YAML", refName)

and same for config map error